native/mod_proxy_cluster/mod_proxy_cluster.c

```diff
@@ -1257,6 +1265,78 @@ static apr_status_t read_node_worker(int id, nodeinfo_t **node, proxy_worker *wo
     return APR_SUCCESS;
 }
 
+static void* APR_THREAD_FUNC check_proxy_worker(apr_thread_t *thread, void *data)
+{
+    apr_status_t rv;
+    char sport[7];
+    char *url;
+    apr_pool_t *rrp;
+    request_rec *rnew;
+
+    watchdog_thread_args_t *targs = (watchdog_thread_args_t *)data;
+    proxy_worker *worker = targs->worker;
+    apr_pool_t *pool = targs->pool;
+    proxy_server_conf *conf = targs->conf;
+    nodeinfo_t *ou = targs->ou;
+    server_rec *server = targs->server;
+    apr_time_t now = targs->now;
+    int id = targs->id;
+
+    apr_snprintf(sport, sizeof(sport), "%d", worker->s->port);
+
+    if (strchr(worker->s->hostname, ':') != NULL)
+        url = apr_pstrcat(pool, worker->s->scheme, "://[", worker->s->hostname, "]:", sport, "/", NULL);
+    else
+        url = apr_pstrcat(pool, worker->s->scheme, "://", worker->s->hostname,  ":", sport, "/", NULL);
+
+    apr_pool_create(&rrp, pool);
+    apr_pool_tag(rrp, "subrequest");
+    rnew = apr_pcalloc(rrp, sizeof(request_rec));
+    rnew->pool = rrp;
+    /* we need only those ones */
+    rnew->server = server;
+    rnew->connection = apr_pcalloc(rrp, sizeof(conn_rec));
+    rnew->connection->log_id = "-";
+    rnew->connection->conn_config = ap_create_conn_config(rrp);
+    rnew->log_id = "-";
+    rnew->useragent_addr = apr_pcalloc(rrp, sizeof(apr_sockaddr_t));
+    rnew->per_dir_config = server->lookup_defaults;
+    rnew->notes = apr_table_make(rnew->pool, 1);
+    rnew->method = "PING";
+    rnew->uri = "/";
+    rnew->headers_in = apr_table_make(rnew->pool, 1);
+    rv = proxy_cluster_try_pingpong(rnew, worker, url, conf, ou->mess.ping, ou->mess.timeout);
+
+    /* We have checked the worker... check if we were told to stop */
+    if (child_stopping) {
+        return APR_SUCCESS;
+    } 
+
+    ap_assert(node_storage->lock_nodes() == APR_SUCCESS);
+    if (read_node_worker(id, &ou, worker) != APR_SUCCESS) {
+        /* the node is gone or something like that */
+        node_storage->unlock_nodes();
+        return APR_SUCCESS;
+    }
+
+    if (rv != APR_SUCCESS) {
+        /* We can't reach the node: XXX changing ou->mess.updatetimelb here ??? */
+        /* XXX if this is a timeout, we might have a outdated list of nodes!!! */
+        worker->s->status |= PROXY_WORKER_IN_ERROR;
+        ou->mess.num_failure_idle++;
+        if (ou->mess.num_failure_idle > 60) {
+            /* Failing for 5 minutes: time to mark it removed */
+            ou->mess.remove = 1;
+            ou->updatetime = now;
+        } 
+    } else {
+        ou->mess.num_failure_idle = 0;
+    }
+
+    node_storage->unlock_nodes();
+    return APR_SUCCESS;
+}
+
```

Comment:
I miss deallocation of the allocated pool. Is it because we rely on the (parent) pool deallocation?

Reply:
yes we rely on the parent pool deallocation.

Reply:
In this case, the parent pool is the `server->process->pool`. How often does it get deallocated, couldn't it be a problem? (I don't know.)

Reply:
OK so you are right the server->process->pool is not deallocated for whole life time of the child process...